[WIP] To Add an option to create multiple Standby MDS nodes in CephFilesystem CRD

xenolinux opened the pull request:
**Description of your changes:**
 Added an option to create multiple Standby MDS nodes in CephFilesystem CRD

**Which issue is resolved by this Pull Request:**
Resolves rook#4068

**Checklist:**

- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

[skip ci]

xenolinux commented:
@leseb Working on it. I will push the changes in some time.

xenolinux commented on pkg/operator/ceph/file/mds/mds.go:
```diff
@@ -120,6 +120,9 @@ func (c *Cluster) Start() error {
 	// Always create double the number of metadata servers to have standby mdses available
 	replicas := c.fs.Spec.MetadataServer.ActiveCount * 2
+	if c.fs.Spec.MetadataServer.ActiveStandByCount > 0 {
+		replicas = c.fs.Spec.MetadataServer.ActiveStandByCount
```
Yes. Added the changes. Thanks

xenolinux replied:
@travisn Working on the above suggestions.

leseb commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -289,6 +289,9 @@ type MetadataServerSpec struct {
 	ActiveCount int32 `json:"activeCount"`
 
+	// The number of standby metadata servers that are active.
+	ActiveStandbyCount int32 `json:"activestandbycount"`
```
How does this option deal with `ActiveStandby` aren't conflicting with each other?

xenolinux replied:
When testing, I set `activestandbycount` to `activestandbycount: 1` in `cephfilesystem`. One new MDS pod was coming up. So, there were a total 3 MDS pods, 1 active and 2 active standby.

leseb commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -289,6 +289,9 @@ type MetadataServerSpec struct {
 	ActiveCount int32 `json:"activeCount"`
 
+	// The number of standby metadata servers that are active.
+	ActiveStandbyCount int32 `json:"activestandbycount"`
```
int32 seems a bit optimistic, `int16` should be enough, not to say 8...

xenolinux replied:
Okay. I will make it `int16`.

xenolinux commented:
The main changes done: 
1. Added CRD schema validation for this setting similar to the `activeCount` setting.  
2. add this validation in an integration test. 
3. Squashed the commits.
4. Converted ActiveStandbyCount type from int32 to int16. 

Remaining changes: 
1. To add the setting in the documentation.

xenolinux commented on tests/framework/installer/ceph_manifests.go:
```diff
@@ -1868,7 +1868,7 @@ func (m *CephManifestsMaster) GetBlockPoolStorageClass(namespace string, poolNam
 
 // GetFilesystem returns the manifest to create a Rook filesystem resource with the given config.
-func (m *CephManifestsMaster) GetFilesystem(namespace, name string, activeCount int) string {
+func (m *CephManifestsMaster) GetFilesystem(namespace, name string, activeCount int, activeStandbyCount int) string {
```
Yes. I do intend to add.

xenolinux commented on cluster/charts/rook-ceph/templates/resources.yaml:
```diff
@@ -197,6 +197,10 @@ spec:
                   type: integer
                 activeStandby:
                   type: boolean
+                activeStanbyCount:
```
Oh yes. I will correct that.

xenolinux commented on cluster/charts/rook-ceph/templates/resources.yaml:
```diff
@@ -197,6 +197,10 @@ spec:
                 activeStandby:
                   type: boolean
+                activeStandbyCount:
+                  minimum: 1
```
Yes. it is optional.

xenolinux commented on cluster/examples/kubernetes/ceph/filesystem.yaml:
```diff
@@ -28,6 +28,8 @@ spec:
     # Whether each active MDS instance will have an active standby with a warm metadata cache for faster failover.
     # If false, standbys will be available, but will not have a warm cache.
     activeStandby: true
```
ack

xenolinux commented:
@leseb I will update the suggested changes today asap.

xenolinux commented:
/test continuous-integration/jenkins/pr-head

leseb commented on tests/framework/installer/ceph_manifests.go:
```diff
@@ -284,6 +284,10 @@ spec:
+		            activeStandbyCount:
+		              minimum: 0
+		              maximum: 10
+		  					  type: integer
```
indent?

leseb commented on tests/framework/installer/ceph_manifests.go:
```diff
@@ -2042,7 +2050,7 @@ func (m *CephManifestsMaster) GetBlockPoolStorageClass(namespace string, poolNam
 
 // GetFilesystem returns the manifest to create a Rook filesystem resource with the given config.
-func (m *CephManifestsMaster) GetFilesystem(namespace, name string, activeCount int) string {
+func (m *CephManifestsMaster) GetFilesystem(namespace, name string, activeCount int, activeStandbyCount int) string {
```
```suggestion
func (m *CephManifestsMaster) GetFilesystem(namespace, name string, activeCount, activeStandbyCount int) string {
```

xenolinux replied:
@leseb I need to make the same change in `tests/framework/installer/ceph_manifests_v1.1.go` right?

leseb replied:
No.

leseb commented on tests/framework/installer/ceph_manifests_v1.1.go:
```diff
@@ -218,6 +218,10 @@ spec:
                 activeStandby:
                   type: boolean
+					    	activeStandbyCount:
+								  minimum: 0
```
indent?

xenolinux commented:
Working on removing unnecessary spaces my editor is adding again and again.

leseb commented:
@xenolinux please fix your rebase.

xenolinux commented:
@leseb @travisn Opened new PR [https://github.com/rook/rook/pull/5034](https://github.com/rook/rook/pull/5034)
